# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [75]:
import pandas as pd
import numpy as np

In [76]:
df = pd.read_csv('demo01.csv')
display(df.head());df.shape

sexo posse_de_veiculo posse_de_imovel  qtd_filhos            tipo_renda  \
0    M                Y               Y           0               Working   
1    F                N               Y           0  Commercial associate   
2    F                N               Y           0  Commercial associate   
3    M                Y               Y           0               Working   
4    F                Y               N           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   

       idade  tempo_emprego  possui_celular  possui_fone_comercial  \
0  58.832877       3.106849               1                      0   
1  52.356164       8.358904               1                      0   
2  52.356164       8.358904               1                      0   
3  46.224658       2.106849               1                      1   
4  29.230137       3.021918               1                      0   

   possui_fone  possui_email  qt_pessoas_residencia    mau  
0            0             0                    2.0  False  
1            1             1                    1.0  False  
2            1             1                    1.0  False  
3            1             1                    2.0  False  
4            0             0                    2.0  False

(16650, 16)

In [77]:
print(df.info());print(f"-----------\n{df.dtypes}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  float64
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  float64
 15  ma

In [78]:
df.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

In [79]:
print(df.mau.value_counts())
print('=-'*20)
print(df.shape[0])
print('=-'*20)
print(round((df.mau.value_counts()/df.shape[0])*100,2))

False    16260
True       390
Name: mau, dtype: int64
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
16650
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
False    97.66
True      2.34
Name: mau, dtype: float64


#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável. 
    Dica: 
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis, 
        3. conte o número de categorias dessa variável 
        4. acumule essa informação de 3. na lista que você criou em 1. 
        5. No final, essa lista pode ser atribuída à nossa variável.
3. Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}.

*Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.*

In [80]:
df_variav = pd.DataFrame(df.dtypes).reset_index()
df_variav.rename(columns={'index':'variavel',0:'type'},inplace=True)

Adicione uma coluna nesse dataframe chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável. 

In [81]:
new_column = list()

for coluna in df.columns.to_list():
    new_column.append(df[coluna].nunique())

df_variav['qtd_categorias'] = new_column
df_variav

variavel     type  qtd_categorias
0                    sexo   object               2
1        posse_de_veiculo   object               2
2         posse_de_imovel   object               2
3              qtd_filhos    int64               8
4              tipo_renda   object               5
5                educacao   object               5
6            estado_civil   object               5
7         tipo_residencia   object               6
8                   idade  float64            5298
9           tempo_emprego  float64            3005
10         possui_celular    int64               1
11  possui_fone_comercial    int64               2
12            possui_fone    int64               2
13           possui_email    int64               2
14  qt_pessoas_residencia  float64               9
15                    mau     bool               2

Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}.

In [82]:
#inicialmente tratando essas variáves contínuas. Pq não podem ser do tipo 'Float'
df.idade = df.idade.astype(int)
df.qt_pessoas_residencia = df.qt_pessoas_residencia.astype(int)
df.mau = df.mau.astype(int)
df.info();df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  int32  
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  int32  
 15  ma

sexo posse_de_veiculo posse_de_imovel  qtd_filhos            tipo_renda  \
0    M                Y               Y           0               Working   
1    F                N               Y           0  Commercial associate   
2    F                N               Y           0  Commercial associate   
3    M                Y               Y           0               Working   
4    F                Y               N           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   

   idade  tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0     58       3.106849               1                      0            0   
1     52       8.358904               1                      0            1   
2     52       8.358904               1                      0            1   
3     46       2.106849               1                      1            1   
4     29       3.021918               1                      0            0   

   possui_email  qt_pessoas_residencia  mau  
0             0                      2    0  
1             1                      1    0  
2             1                      1    0  
3             1                      2    0  
4             0                      2    0

In [83]:
df['posse_de_veiculo'] = df['posse_de_veiculo'].map({'Y':1,'N':0})
df['posse_de_imovel'] = df['posse_de_imovel'].map({'Y':1,'N':0})
df['sexo'] = df['sexo'].map({'M':0,'F':1})

In [84]:
df_base = pd.get_dummies(df,['tipo_renda','estado_civil','educacao','tipo_residencia'])
display(df.head())
display(df_base.head())

sexo  posse_de_veiculo  posse_de_imovel  qtd_filhos            tipo_renda  \
0     0                 1                1           0               Working   
1     1                 0                1           0  Commercial associate   
2     1                 0                1           0  Commercial associate   
3     0                 1                1           0               Working   
4     1                 1                0           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   

   idade  tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0     58       3.106849               1                      0            0   
1     52       8.358904               1                      0            1   
2     52       8.358904               1                      0            1   
3     46       2.106849               1                      1            1   
4     29       3.021918               1                      0            0   

   possui_email  qt_pessoas_residencia  mau  
0             0                      2    0  
1             1                      1    0  
2             1                      1    0  
3             1                      2    0  
4             0                      2    0

sexo  posse_de_veiculo  posse_de_imovel  qtd_filhos  idade  tempo_emprego  \
0     0                 1                1           0     58       3.106849   
1     1                 0                1           0     52       8.358904   
2     1                 0                1           0     52       8.358904   
3     0                 1                1           0     46       2.106849   
4     1                 1                0           0     29       3.021918   

   possui_celular  possui_fone_comercial  possui_fone  possui_email  ...  \
0               1                      0            0             0  ...   
1               1                      0            1             1  ...   
2               1                      0            1             1  ...   
3               1                      1            1             1  ...   
4               1                      0            0             0  ...   

   educacao_Married  educacao_Separated  educacao_Single / not married  \
0                 1                   0                              0   
1                 0                   0                              1   
2                 0                   0                              1   
3                 1                   0                              0   
4                 1                   0                              0   

   educacao_Widow  tipo_residencia_Co-op apartment  \
0               0                                0   
1               0                                0   
2               0                                0   
3               0                                0   
4               0                                0   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                                  1                                    0   
1                                  1                                    0   
2                                  1                                    0   
3                                  1                                    0   
4                                  1                                    0   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                                 0                                 0   
1                                 0                                 0   
2                                 0                                 0   
3                                 0                                 0   
4                                 0                                 0   

   tipo_residencia_With parents  
0                             0  
1                             0  
2                             0  
3                             0  
4                             0  

[5 rows x 33 columns]

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [85]:
#acredito que fiz isso no item anterior
display(df_base)
df_base.info()

sexo  posse_de_veiculo  posse_de_imovel  qtd_filhos  idade  \
0         0                 1                1           0     58   
1         1                 0                1           0     52   
2         1                 0                1           0     52   
3         0                 1                1           0     46   
4         1                 1                0           0     29   
...     ...               ...              ...         ...    ...   
16645     1                 0                1           0     54   
16646     1                 0                1           0     43   
16647     0                 1                1           0     30   
16648     0                 1                1           0     30   
16649     1                 0                1           0     33   

       tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0           3.106849               1                      0            0   
1           8.358904               1                      0            1   
2           8.358904               1                      0            1   
3           2.106849               1                      1            1   
4           3.021918               1                      0            0   
...              ...             ...                    ...          ...   
16645       9.884932               1                      0            0   
16646       7.380822               1                      1            1   
16647       9.800000               1                      1            0   
16648       9.800000               1                      1            0   
16649       3.630137               1                      0            1   

       possui_email  ...  educacao_Married  educacao_Separated  \
0                 0  ...                 1                   0   
1                 1  ...                 0                   0   
2                 1  ...                 0                   0   
3                 1  ...                 1                   0   
4                 0  ...                 1                   0   
...             ...  ...               ...                 ...   
16645             0  ...                 0                   0   
16646             0  ...                 1                   0   
16647             0  ...                 1                   0   
16648             0  ...                 1                   0   
16649             1  ...                 1                   0   

       educacao_Single / not married  educacao_Widow  \
0                                  0               0   
1                                  1               0   
2                                  1               0   
3                                  0               0   
4                                  0               0   
...                              ...             ...   
16645                              0               0   
16646                              0               0   
16647                              0               0   
16648                              0               0   
16649                              0               0   

       tipo_residencia_Co-op apartment  tipo_residencia_House / apartment  \
0                                    0                                  1   
1                                    0                                  1   
2                                    0                                  1   
3                                    0                                  1   
4                                    0                                  1   
...                                ...                                ...   
16645                                0                                  1   
16646                                0                                  1   
16647                                0                                  1   
16648                        

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 33 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   sexo                                        16650 non-null  int64  
 1   posse_de_veiculo                            16650 non-null  int64  
 2   posse_de_imovel                             16650 non-null  int64  
 3   qtd_filhos                                  16650 non-null  int64  
 4   idade                                       16650 non-null  int32  
 5   tempo_emprego                               16650 non-null  float64
 6   possui_celular                              16650 non-null  int64  
 7   possui_fone_comercial                       16650 non-null  int64  
 8   possui_fone                                 16650 non-null  int64  
 9   possui_email                                16650 non-null  int64  
 10  qt_pessoas

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [86]:
tab = pd.crosstab(df_base['possui_email'],df_base['posse_de_veiculo'])
tab.div(tab.sum(axis=1),axis=0)

posse_de_veiculo         0         1
possui_email                        
0                 0.615755  0.384245
1                 0.565541  0.434459

In [87]:
pd.crosstab(df_base['possui_email'],df_base['mau']).div(tab.sum(axis=1),axis=0)

mau                  0         1
possui_email                    
0             0.976269  0.023731
1             0.979730  0.020270

In [88]:
pd.crosstab(df_base['posse_de_veiculo'],df_base['mau']).div(tab.sum(axis=1),axis=0)

mau                      0         1
posse_de_veiculo                    
0                 0.654647  0.016282
1                 4.276351  0.096622

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [89]:
df_base.to_csv('demo01_clean.csv')

A diferença nas proporções de mau = 1 é maior para posse_de_veiculo do que para possui_email. A variável posse_de_veiculo mostra uma variação mais significativa entre as categorias, especialmente a categoria posse_de_veiculo = 1, que tem uma proporção de mau = 1 significativamente maior (9.66%) em comparação com a outra categoria (1.63%).

Portanto, posse_de_veiculo parece ser uma variável mais poderosa para prever a probabilidade de mau = 1, uma vez que a diferença nas proporções de mau = 1 entre suas categorias é mais pronunciada.